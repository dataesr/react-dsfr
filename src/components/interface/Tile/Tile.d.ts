import * as React from 'react';

export type TileClassName = string | Object | any[];

export type TileChildren = React.ReactNode[] | React.ReactNode;

export interface TileProps {
    className?: TileClassName;
    /**
     * Source of the image â€” size is fixed 80x80
     */
    horizontal?: boolean;
    color?: string;
    verticalMedium?: boolean;
    horizontalMedium?: boolean;
    /**
     * One of: node, arrayOf(node), string
     */
    children: TileChildren;
    onClick?: (...args: any[])=>any;
    ariaLabel?: string;
}

declare const Tile: React.FC<TileProps>;

export default Tile;

