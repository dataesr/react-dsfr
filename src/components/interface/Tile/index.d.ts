import type Tile from './Tile';
import type TileBody from './TileBody';
import type TileImage from './TileImage';

export type { Tile, TileBody, TileImage };
