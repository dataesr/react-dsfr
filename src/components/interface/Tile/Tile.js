import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { SCHEMES } from '../../../utils/constants';

import '@gouvfr/dsfr/dist/css/tiles.min.css';

/**
 *
 * @visibleName Tile
 */
const Tile = ({
  horizontal,
  className,
  horizontalMedium,
  verticalMedium,
  children,
  scheme,
}) => {
  const _className = classNames('rf-tile rf-enlarge-link', {
    'rf-tile--horizontal': horizontal,
    'rf-tile--horizontal-md': !horizontal && horizontalMedium,
    'rf-tile--vertical-md': verticalMedium && horizontal,
    [`rf-scheme-${scheme}`]: scheme,
  }, className);
  return <div className={_className}>{children}</div>;
};

Tile.defaultProps = {
  className: '',
  scheme: '',
  horizontal: false,
  verticalMedium: false,
  horizontalMedium: false,
};

Tile.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  /**
   * Source of the image â€” size is fixed 80x80
   */
  horizontal: PropTypes.bool,
  verticalMedium: PropTypes.bool,
  horizontalMedium: PropTypes.bool,
  /**
   * One of: node, arrayOf(node), string
   */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  scheme: PropTypes.oneOf(SCHEMES),
};

export default Tile;
