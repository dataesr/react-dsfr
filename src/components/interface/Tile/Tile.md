## Tile - Tuile

La tuile est un point d’entrée qui redirige les utilisateurs vers des pages de contenu. Elle fait partie d'une collection et n’est jamais présentée de manière isolée.

### Tuile horizontale par défaut et verticale en desktop
```jsx

import { Tile, TileBody, TileImage } from '.';

<Tile horizontal={true} verticalMedium={true}>
    <TileBody
        title="Tile Title"
        description="Tile description"
        linkHref='/myPath'>
    </TileBody>
    <TileImage
        alt="alternative"
        src="https://place-hold.it/80x80">
    </TileImage>
</Tile>
```

### Tuile verticale par défaut et horizontale en desktop
```jsx

import { Tile, TileBody, TileImage } from '.';

<Tile horizontalMedium={true} scheme="soft-orange-dark">
    <TileBody
        title="Tile Title"
        description="Tile description"
        linkHref='/myPath'>
    </TileBody>
    <TileImage
        alt="alternative"
        src="https://place-hold.it/80x80">
    </TileImage>
</Tile>
```


### Usage avec Container, Row, Col

```jsx
import { Tile, TileBody, TileImage } from '.';

import Row from '../../foundation/Grid/Row';
import Container from '../../foundation/Grid/Container';
import Col from '../../foundation/Grid/Col';

<Container>
    <Row>
        <Col>
            <Tile>
                <TileBody
                    titleAs="h5"
                    title="Tile Title"
                    linkHref='/myPath'>
                </TileBody>
                <TileImage
                    alt="alternative"
                    src="https://place-hold.it/80x80">
                </TileImage>
            </Tile>
        </Col>
        <Col>
            <Tile>
                <TileBody
                    titleAs="h5"
                    title="Tile Title"
                    description="Tile description"
                    linkHref='/myPath'>
                </TileBody>
            </Tile>
        </Col>
        <Col>
            <Tile>
                <TileBody
                    titleAs="h5"
                    title="Tile Title"
                    linkHref='/myPath'>
                </TileBody>
                <TileImage
                    alt="alternative"
                    src="https://place-hold.it/80x80"
                    linkHref='/myPath'>
                </TileImage>
            </Tile>
        </Col>
    </Row>
</Container>
```
