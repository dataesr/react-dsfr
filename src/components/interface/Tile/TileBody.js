import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Link from '../Link/index';
import dataAttributes from '../../../utils/data-attributes';

/**
 *
 * @visibleName Corps de la tuile
 */
const TileBody = ({
  className,
  title,
  description,
  titleAs,
  href,
  to,
  linkTitle,
  linkTarget,
  ...remainingProps
}) => {
  const HtmlTag = `${titleAs}`;
  return (
    <div
      className={classNames('fr-tile__body', className)}
      {...dataAttributes(remainingProps)}
    >
      <HtmlTag className="fr-tile__title">
        <Link
          className="fr-tile__link"
          title={linkTitle}
          target={linkTarget || undefined}
          href={href}
          to={to}
        >
          {title}
        </Link>
      </HtmlTag>
      {description && <p className="fr-tile__desc">{description}</p>}
    </div>
  );
};

TileBody.defaultProps = {
  description: '',
  titleAs: 'p',
  linkTitle: '',
  linkTarget: '',
  className: '',
  href: undefined,
  to: undefined,
};

TileBody.propTypes = {
  /**
   * Html tag to render in accordion title.
   */
  titleAs: PropTypes.oneOf(['p', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6']),
  title: PropTypes.string.isRequired,
  description: PropTypes.string,
  linkTarget: PropTypes.string,
  linkTitle: PropTypes.string,
  href: PropTypes.string,
  to: PropTypes.string,
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
};

export default TileBody;
