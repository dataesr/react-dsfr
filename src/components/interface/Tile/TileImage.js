import PropTypes from 'prop-types';
import classNames from 'classnames';

/**
 * La tuile est un point d’entrée qui redirige les utilisateurs vers des pages de contenu.
 * Elle fait partie d'une collection et n’est jamais présentée de manière isolée.
 *
 * @visibleName Tuile horizontale ou verticale
 */
const TileImage = ({
  title,
  src,
  alt,
  className,
}) => (
  <div className={classNames('rf-tile__img-wrap', className)}>
    <img
      src={src}
      title={title}
      alt={alt}
    />
  </div>
);

TileImage.defaultProps = {
  className: '',
  title: '',
};

TileImage.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  title: PropTypes.string,
  alt: PropTypes.string.isRequired,
  /**
   * Source of the image — size is fixed 80x80
   */
  src: PropTypes.string.isRequired,
};

export default TileImage;
