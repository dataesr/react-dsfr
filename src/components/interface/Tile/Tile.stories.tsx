import { ComponentStory } from '@storybook/react';

import Tile from './Tile';
import TileBody from './TileBody';
import TileImage from './TileImage';

// https://www.systeme-de-design.gouv.fr/elements-d-interface/composants/tuile

export default {
  title: 'Composants/Tuile - Tile',
  component: Tile,
  subcomponents: { TileBody, TileImage },
  parameters: {
    docs: {
      description: {
        component: 'La tuile est un raccourci ou point d\'entrée qui redirige les utilisateurs vers des pages de contenu. Elle fait généralement partie d\'une collection ou liste de tuiles similaires. La tuile n\'est jamais présentée de manière isolée.',
      },
    },
  },
};

const Template: ComponentStory<typeof Tile> = (args) => (
  <Tile {...args} />
);

export const SimpleTile = Template.bind({});
SimpleTile.args = {
  children: [
    <TileBody title="Titre MD bold">
      Description
    </TileBody>,
  ],
};
SimpleTile.storyName = 'Tuile simple';

export const SimpleTileWithImage = Template.bind({});
SimpleTileWithImage.args = {
  children: [
    <TileBody title="Titre MD bold">
      Tuile avec image
    </TileBody>,
    <TileImage alt="" src="https://picsum.photos/200/300" />,
  ],
};
SimpleTileWithImage.storyName = 'Tuile simple avec image';

export const HorizontalTile = Template.bind({});
HorizontalTile.args = {
  children: [
    <TileBody title="Titre MD bold">
      Tuile horizontale
    </TileBody>,
  ],
  horizontal: true,
};
HorizontalTile.storyName = 'Tuile horizontale';

export const HorizontalMDTile = Template.bind({});
HorizontalMDTile.args = {
  children: [
    <TileBody title="Titre MD bold">
      Tuile horizontale MD
    </TileBody>,
  ],
  horizontalMedium: true,
};
HorizontalMDTile.storyName = 'Tuile horizontale MD';

export const VerticalMDTile = Template.bind({});
VerticalMDTile.args = {
  children: [
    <TileBody title="Titre MD bold">
      Tuile verticale MD
    </TileBody>,
  ],
  verticalMedium: true,
};
VerticalMDTile.storyName = 'Tuile verticale MD';
