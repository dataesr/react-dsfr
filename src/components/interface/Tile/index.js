import Tile from './Tile';
import TileBody from './TileBody';
import TileImage from './TileImage';

export {
  Tile,
  TileBody,
  TileImage,
};
