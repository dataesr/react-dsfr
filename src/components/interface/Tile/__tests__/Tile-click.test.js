import Enzyme, { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Tile from '../Tile';
import TileBody from '../TileBody';
import TileImage from '../TileImage';

Enzyme.configure({ adapter: new Adapter() });

const mockSetonClickTile = jest.fn();

const initialProps = {
  horizontal: true,
  verticalMedium: true,
  title: 'title',
  linkTitle: 'linkTitle',
  linkHref: '/path',
  linkTarget: 'linkTarget',
  alt: '',
  src: '/src.png',
  description: 'description',
  className: 'my-class',
};

describe('<Tile />', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = () => shallow(
      <Tile
        onClick={mockSetonClickTile}
        horizontal={initialProps.horizontal}
        className={initialProps.className}
        data-testid="tile"
      >
        <TileBody
          title={initialProps.title}
          description={initialProps.description}
          linkTitle={initialProps.linkTitle}
          linkTarget={initialProps.linkTarget}
          data-test-id="tilebody"
        >
          <p>Du texte aussi</p>
        </TileBody>
        <TileImage
          alt={initialProps.alt}
          src={initialProps.src}
          data-id="titleimage"
        />
      </Tile>,
    );
  });

  it('renders correctly', () => {
    const component = renderer
      .create(
        <Tile
          onClick={mockSetonClickTile}
          horizontal={initialProps.horizontal}
          className={initialProps.className}
          data-testid="tile"
        >
          <TileBody
            title={initialProps.title}
            description={initialProps.description}
            linkTitle={initialProps.linkTitle}
            linkTarget={initialProps.linkTarget}
            data-test-id="tilebody"
          >
            <p>Du texte aussi</p>
          </TileBody>
          <TileImage
            alt={initialProps.alt}
            src={initialProps.src}
            data-id="titleimage"
          />
        </Tile>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('on onClickTile', () => {
    const component = wrapper({ ...initialProps });
    component.find('[role="button"]').simulate('click', { preventDefault: () => {} });
    expect(mockSetonClickTile).toHaveBeenCalledTimes(1);
  });
});
