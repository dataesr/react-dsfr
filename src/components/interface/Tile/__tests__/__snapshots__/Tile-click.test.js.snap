// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tile /> renders correctly 1`] = `
<div
  className="fr-tile fr-enlarge-link fr-tile--horizontal my-class"
  data-testid="tile"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex="0"
>
  <div
    className="fr-tile__body"
    data-test-id="tilebody"
  >
    <h3
      className="fr-tile__title"
    >
      <a
        aria-disabled={true}
        className="fr-tile__link fr-link--md"
        role="link"
        target="linkTarget"
        title="linkTitle"
      >
        title
      </a>
    </h3>
    <p
      className="fr-tile__desc"
    >
      description
    </p>
    <p>
      Du texte aussi
    </p>
  </div>
  <div
    className="fr-tile__img"
    data-id="titleimage"
  >
    <img
      alt=""
      src="/src.png"
      title=""
    />
  </div>
</div>
`;
